crates/ruff/src/rules/flake8_bugbear/rules/reuse_of_groupby_generator.rs

```diff
@@ -0,0 +1,143 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::types::Range;
+use ruff_python_ast::visitor::{self, Visitor};
+use rustpython_parser::ast::{Expr, ExprKind, Stmt, StmtKind};
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for multiple usage of the generator returned from
+/// `itertools.groupby()`.
+///
+/// ## Why is it bad?
+/// Using the generator more than once will do nothing on the second usage.
+/// If that data is needed later, it should be stored as a list.
+///
+/// ## Examples:
+/// ```python
+/// import itertools
+///
+/// for name, group in itertools.groupby(data):
+///     for _ in range(5):
+///         do_something_with_the_group(group)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// import itertools
+///
+/// for name, group in itertools.groupby(data):
+///     values = list(group)
+///     for _ in range(5):
+///         do_something_with_the_group(values)
+/// ```
+#[violation]
+pub struct ReuseOfGroupbyGenerator;
+
+impl Violation for ReuseOfGroupbyGenerator {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Using the generator returned from `itertools.groupby()` more than once will do nothing on the second usage. Save the result to a list, if the result is needed multiple times.")
+    }
+}
+
+/// A [`Visitor`] that collects all the usage of `group_name` in the body of
+/// a `for` loop.
+struct GroupNameFinder<'a> {
+    /// Variable name for the group.
+    group_name: &'a str,
+    /// Number of times the `group_name` variable was seen during the visit.
+    usage_count: u8,
+    /// A flag indicating that the visitor is inside a nested `for` loop.
+    nested: bool,
+    exprs: Vec<&'a Expr>,
+}
+
+impl<'a> GroupNameFinder<'a> {
+    fn new(group_name: &'a str) -> Self {
+        GroupNameFinder {
+            group_name,
+            usage_count: 0,
+            nested: false,
+            exprs: Vec::new(),
+        }
+    }
+}
+
+impl<'a, 'b> Visitor<'b> for GroupNameFinder<'a>
+where
+    'b: 'a,
+{
+    fn visit_stmt(&mut self, stmt: &'a Stmt) {
+        match &stmt.node {
+            StmtKind::For { .. } => {
+                self.nested = true;
+                visitor::walk_stmt(self, stmt);
+                self.nested = false;
+            }
+            _ => visitor::walk_stmt(self, stmt),
+        }
+    }
+
+    fn visit_expr(&mut self, expr: &'a Expr) {
+        if let ExprKind::Name { id, .. } = &expr.node {
+            if id == self.group_name {
```

Comment:
Not sure how to handle it but this analysis will yield false-positives if `id` is re-bound`. For example, because a new value is assigned, the visitor enters a function that has an argument with the same name, etc.

Reply:
Right! Variable re-assignment inside the loop might yield false-positives. From the top of my mind, I'm thinking of skipping the checks during the scope of the variable assignment. I'm not sure about the scope determination part but something along the lines of `self.nested` should work.